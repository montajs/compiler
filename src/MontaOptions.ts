export interface MontaOptions {
	/**
	 * Directory containing the Monta templates
	 *
	 * When rendering a template file with a relative path, it will be looked up
	 * in this directory. Relative paths will be resolved from the current
	 * working directory.
	 *
	 * @default ./views
	 */
	templateRoot : string;

	/**
	 * Set to false to disable caching templates
	 *
	 * @default process.env.NODE_ENV !== 'development'
	 */
	cache : boolean;

	/**
	 * Maximum number of templates to keep in cache
	 *
	 * Whenever a Monta template file is rendered, the parsed template is cached
	 * to speed up subsequent rendering. This may consume several megabytes of
	 * working memory, depending on the number of templates. Set this value to
	 * limit the number of templates in the cache.
	 *
	 * It is recommended to keep this set to the default value of Infinity, unless
	 * you experience memory problems.
	 *
	 * @default Infinity
	 */
	cacheLimit : number;

	/**
	 * Set to true to enable debug logging
	 *
	 * Do not use in production, this option logs a lot of unnecessary detail.
	 *
	 * @default false
	 */
	debug : boolean;
}
