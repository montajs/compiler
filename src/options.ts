import path from 'path';

import { MontaOptions } from './MontaOptions';

export let options : MontaOptions = {
	templateRoot: process.cwd(),

	cache: process.env.NODE_ENV !== 'development',
	cacheLimit: Infinity,

	debug: false,
};

export function configure(config : Partial<MontaOptions>) : void {
	if (config.templateRoot !== undefined) {
		options.templateRoot = path.resolve(config.templateRoot);
	}

	if (config.cache !== undefined) {
		options.cache = Boolean(config.cache);
	}

	if (config.cacheLimit !== undefined && Number.isInteger(config.cacheLimit)) {
		options.cacheLimit = config.cacheLimit;
	}

	if (config.debug !== undefined) {
		options.debug = Boolean(config.debug);
	}
}
