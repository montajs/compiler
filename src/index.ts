import { Template } from './Template';

export { registerParserFunction, registerFunction } from './functions/functions';
export { configure } from './options';
export { parse } from './parser';
export { Template } from './Template';

export async function render(filename : string, data : Record<string, any> = {}) : Promise<string> {
	let template = await Template.create(filename);
	return template.render(data);
}
