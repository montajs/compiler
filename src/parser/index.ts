import { readFile } from 'fs-extra';

import { ParseContext } from './ParseContext';
import { scan } from './scan';
import { parseSection } from './parse';
import { GroupNode } from './nodes/GroupNode';
import { RenderFn } from '../Template';
import { Logger } from '../utils/Logger';

/**
 * Parses the source code from a given parser context
 *
 * @param parseContext - The parser context
 */
export async function parse(parseContext : ParseContext) : Promise<RenderFn> {
	Logger.group('READ');
	Logger.info('| filename: %s', parseContext.file);
	let code = await readFile(parseContext.file, 'utf-8');
	Logger.info('| length: %d', code.length);
	Logger.groupEnd();

	Logger.group('SCAN');
	let sections = scan(code);
	Logger.groupEnd();

	Logger.group('PARSE');
	Logger.info('| sections: %d', sections.length);
	let rootNode = new GroupNode();

	for (let section of sections) {
		rootNode.children.push(...parseSection(section));
	}
	Logger.groupEnd();

	Logger.group('PREPARE');
	let functions = await rootNode.createRenderFunction(parseContext);
	Logger.groupEnd();

	return functions;
}
