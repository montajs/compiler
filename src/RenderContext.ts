import { TokenType } from './parser/TokenType';
import { Token } from './parser/Token';
import { RenderFn } from './Template';
import { Logger } from './utils/Logger';

export type Renderable = string | number | boolean | Record<string, any> | undefined | RenderFn;

export class RenderContext {

	/** Full path to the file that is currently being parsed */
	public file : string;

	private parent ? : RenderContext;
	private root : RenderContext;

	private variables : Record<string, any>;
	private data = new Map<string, any>();

	public constructor(file : string, variables : Record<string, any> = {}) {
		this.file = file;
		this.variables = variables;

		this.root = this;
	}

	public getData<T>(namespace : string, key : string) : T | undefined {
		return this.data.get(`${ namespace }:${ key }`) as T;
	}

	public setData<T>(namespace : string, key : string, value : T) : void {
		this.data.set(`${ namespace }:${ key }`, value);
	}

	public getVariable(path : string) : any | undefined {
		if (path === '.' || path === 'this') {
			return this.variables;
		}

		let data : any = this.variables;

		let keys = path.split('.');

		if (keys[0] === 'this' || keys[0] === '') {
			keys.shift();
		}

		let scopeChanged = false;

		if (keys[0] === '$parent') {
			if (!this.parent) throw new Error('Current scope has no parent scope');

			scopeChanged = true;
			data = this.parent.data;
			keys.shift();
		}

		if (keys[0] === '$root') {
			scopeChanged = true;
			data = this.root.data;
			keys.shift();
		}

		if (!scopeChanged && typeof this.data !== 'object') {
			throw new Error(`Cannot get property '${ path }' of primitive value`);
		}

		while (keys.length > 0) {
			let key = keys.shift() as string;

			if (data === undefined) {
				break;
			}

			if (typeof data !== 'object') {
				throw new TypeError(`Cannot get property '${ key }' of primitive value`);
			}

			data = (data as Record<string, any>)[key];
		}

		if (data === undefined) {
			return this.parent?.getVariable(path); // eslint-disable-line @typescript-eslint/no-unsafe-return
		}

		return data; // eslint-disable-line @typescript-eslint/no-unsafe-return
	}

	public setVariable(key : string, value : any) : void {
		this.variables[key] = value; // TODO handle '.' paths
	}

	public getValue(ident : Token | undefined) : any | undefined {
		if (ident === undefined) {
			throw new Error('undefined ident');
		}

		let type = typeof ident;
		if (type === 'string' || type === 'number' || type === 'boolean') {
			return ident;
		}

		if (ident.type === TokenType.Literal) {
			return ident.value;
		}

		return this.getVariable(ident.value.toString()); // eslint-disable-line @typescript-eslint/no-unsafe-return
	}

	public createChildContext(file ? : string) : RenderContext {
		let context = new RenderContext(file ?? this.file, this.variables);
		context.data = this.data;
		context.parent = this;
		return context;
	}

	public async render(render : Renderable[]) : Promise<string> {
		let rendered : Renderable[] = render;
		let isValid = false;

		while (!isValid) {
			isValid = true;

			rendered = await Promise.all(rendered.flat().map(child => {
				let type = typeof child;

				if (type === 'string') {
					return child;
				}

				if (child instanceof Function) {
					isValid = false;
					return child(this);
				}

				if (child === undefined) {
					return 'undefined';
				}

				if (type === 'number') {
					return child.toString();
				}

				if (type === 'boolean') {
					return child ? 'true' : 'false';
				}

				if (type === 'object') {
					return JSON.stringify(child, undefined, 2);
				}

				Logger.error('Not renderable:', child);
				throw new Error(`Not renderable`);
			}));
		}

		return rendered.join('');
	}
}
