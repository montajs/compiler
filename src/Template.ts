import { RenderContext } from './RenderContext';
import { ParseContext } from './parser/ParseContext';
import { parse } from './parser';

export type RenderFnOutput = string | number | boolean | RenderFn;
export type RenderFn = (context : RenderContext, input ? : any) => RenderFnOutput | RenderFnOutput[] | Promise<RenderFnOutput> | Promise<RenderFnOutput[]>;

export class Template {

	public static async create(filename : string) : Promise<Template> {
		let parseContext = new ParseContext(filename);
		let renderFn = await parse(parseContext);

		return new Template(parseContext, renderFn);
	}

	/** Source file of this template */
	public get file() : string {
		return this.parseContext.file;
	}

	private readonly parseContext : ParseContext;

	/** The root node of the template */
	private readonly renderFn : RenderFn;

	public constructor(parseContext : ParseContext, renderFn : RenderFn) {
		this.parseContext = parseContext;
		this.renderFn = renderFn;
	}

	/**
	 * Renders the template to a string
	 *
	 * @param data - The data to replace
	 */
	public async render(data : Record<string, any>) : Promise<string> {
		let renderContext = new RenderContext(this.file, data);
		return renderContext.render([this.renderFn]);
	}
}
