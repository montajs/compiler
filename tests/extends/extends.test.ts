import { configure, render } from '../../src';

beforeEach(() => {
	configure({
		templateRoot: __dirname,
	});
});

test('base', async () => {
	let rendered = await render('extends-base');
	expect(rendered).toMatchSnapshot();
});

test('nested', async () => {
	let rendered = await render('extends-nested');
	expect(rendered).toMatchSnapshot();
});

test('default block', async () => {
	let rendered = await render('extends-default-block');
	expect(rendered).toMatchSnapshot();
});

test('empty block', async () => {
	let rendered = await render('extends-empty-block');
	expect(rendered).toMatchSnapshot();
});

test('relative', async () => {
	let rendered = await render('extends-relative');
	expect(rendered).toMatchSnapshot();
});
