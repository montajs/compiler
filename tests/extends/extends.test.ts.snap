// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base 1`] = `
"

<html>
<head>
	<title>test</title>
	
</head>
<body>

	<p>I extend a base template</p>

</body>
</html>



"
`;

exports[`default block 1`] = `
"
	<p>Default content</p>


"
`;

exports[`empty block 1`] = `
"

"
`;

exports[`nested 1`] = `
"

<html>
<head>
	<title>nested</title>
	
	<link rel=\\"stylesheet\\" href=\\"style.css\\">

</head>
<body>

	<p>I extend a nested base template</p>

</body>
</html>






"
`;

exports[`relative 1`] = `
"
	<p>Default content</p>



"
`;
